rfcs/0008-storage-ref.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: Storage References
+- Start Date: 2023-02-10)
+- RFC PR: [FuelLabs/sway-rfcs#0008](https://github.com/FuelLabs/sway-rfcs/pull/023)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/3043)
+
+# Summary
+
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+# Motivation
+
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A storage reference stores a `b256` key that describes a storage slot:
+
+```rust
+pub struct StorageRef<T> {
+    key: b256,
+}
+```
+
+We can implement `store` and `get` as follows:
+
+```rust
+impl<T> StorageRef<T> {
+    #[storage(read)]
+    pub fn get(self) -> Option<T> {
+        std::storage::get(self.key)
+    }
+
+    #[storage(write)]
+    pub fn store(self, other: T) {
+        std::storage::store(self.key, other)
+    }
+}
+```
+
+One can create a `StorageRef` manually (it's really just a wrapper around a `b256` key) or via a call to `as_storage_ref` from a new trait called `AsStorageRef`:
+
+```rust
+pub trait AsStorageRef<T> {
+    fn as_storage_ref(self) -> StorageRef<T>;
+}
+```
+
+Calling `as_storage_ref` is only valid for a storage accesses. A storage access looks like `storage.a.b.c...`. For example:
+
+```rust
+struct MyStruct {
+    x: u64,
+    y: b256,
+}
+
+storage {
+    s: MyStruct, 
+}
+
+fn foo() {
+    let x_ref: StorageRef<u64> = storage.s.x.as_storage_ref();
+    let y_ref: StorageRef<b56> = storage.s.y.as_storage_ref();
+    let s_ref: StorageRef<MyStruct> = storage.s.as_storage_ref();
```

Comment:
Does a storage access have a valid type at all *without* having called `as_storage_ref()`?

If not, could we perhaps assume that any field access on storage (i.e. `storage.x.y.z`) is always a `StorageRef<T>`?

Hmm I guess this might require some compiler magic to enable the access of the field `y` on the `StorageRef<T>` resulting from `storage.x` and so on...

Reply:
The type of `storage.x.y.z` is just the same type as `z`. The problem with this is that calling `as_storage_ref` will only be valid if called directly using `storage.x.y.z.as_storage_ref()` but not valid if called as below:
```
let z = storage.x.y.z;
let z_ref = z.as_storage_ref()`
```
despite `z` having the same type as `storage.x.y.z`.

Another reason why making `storage.x.y.z` of type `StorageRef` might be a cleaner approach, though I'm reluctant to make this change because it complicates the syntax for getting and updating simple storage variables.

Reply:
> The type of storage.x.y.z is just the same type as z.

I realize this is the case today, but are we sure we want to continue to allow this?

E.g. if the field  `z` has the type `Z` and `Z` has some method `foo` that takes `ref mut self`, what happens when you do `storage.x.y.z.foo()`? Does it update storage at the field `z`? The method implementation for the type `Z` would behave differently when the type is used in storage, vs when the method is called on an instance of `Z` that is *not* in storage right? Could the fact that `Z` may have a different layout (whether used in the context of storage or not) not have an impact on the behaviour of the method?

It seems like, for any storage access, we might want to *require* that it's through the lens of some sort of storage reference?

Reply:
No longer relevant since `storage.x.y.z` now returns the storage handler itself.